### PR TITLE
Import transaction; export wallet as pdf; rbf & cpfp fix

### PR DESCRIPTION
**Pull request by yzhurko**

### CODE REVIEW

File: GUI/QtQuick/PaperBackupWriter.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+   , qr_(qr)
+{
+}
+
+bool WalletBackupPdfWriter::write(const QString &fileName)
+{
+   QFile f(fileName);
+   bool success = f.open(QIODevice::WriteOnly);
+
+   if (!success)
```
Comment (sergey-chernikov) [inline GUI/QtQuick/PaperBackupWriter.cpp]: brackets missing

File: GUI/QtQuick/PaperBackupWriter.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
+/*
+
+***********************************************************************************
+* Copyright (C) 2018 - 2020, BlockSettle AB
```
Comment (sergey-chernikov) [inline GUI/QtQuick/PaperBackupWriter.cpp]: Copyright year

File: GUI/QtQuick/PaperBackupWriter.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+   QFont italic = font;
+   italic.setItalic(true);
+
+   QFont bold = font;
+   bold.setBold(true);
+
+   QStaticText wIdDesc(QLatin1String("Wallet ID"));
+   QStaticText keyLine1Desc(QLatin1String("Line 1"));
+   QStaticText keyLine2Desc(QLatin1String("Line 2"));
```
Comment (sergey-chernikov) [inline GUI/QtQuick/PaperBackupWriter.cpp]: Better to have them translatable.

File: GUI/QtQuick/PaperBackupWriter.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// WalletBackupPdfWriter
+//
+
+//! Writer of PDF backup of wallet.
+class WalletBackupPdfWriter final
+{
+public:
+   static constexpr int kResolution = 1200;
+   static constexpr double kTotalWidthInches = 8.27;
+   static constexpr double kTotalHeightInches = 11.69;
```
Comment (sergey-chernikov) [inline GUI/QtQuick/PaperBackupWriter.h:27]: These are likely the sizes of the Letter format, not A4?

File: GUI/QtQuick/PaperBackupWriter.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+public:
+   WalletBackupPdfWriter(const QString &walletId
+      , const QString &keyLine1, const QString &keyLine2
+      , const QPixmap &qr);
+
+   //! Write backup to PDF.
+   bool write(const QString &fileName);
+
+   //! Generate resulted PDF preview pixmap. Height will be calculated using page aspect ratio.
```
Comment (sergey-chernikov) [inline GUI/QtQuick/PaperBackupWriter.h]: "resulting"
what's the meaning of `!` in `//!`?

File: GUI/QtQuick/QtQuickAdapter.cpp [hunk 6]
```diff
@@ -2592,3 +2593,36 @@ bool QtQuickAdapter::broadcastSignedTX(const QUrl& path)
...
    return true;
 }
+
+bool QtQuickAdapter::isRequestReadyToSend(QTXSignRequest* request)
+{
+    const auto& walletIds = request->txReq().walletIds;
+    for (const auto& id : walletIds) {
+        bool isInWallets = false;
+        for (const auto [walletId, wallet] : hdWallets_) {
+            if (wallet.hasLeaf(id) && !wallet.watchOnly) {
```
Comment (sergey-chernikov) [inline GUI/QtQuick/QtQuickAdapter.cpp]: IMO, you should also check that `id` is `walletId` of HD Wallet.

File: GUI/QtQuick/TxListModel.cpp [hunk 7]
```diff
@@ -166,18 +171,6 @@ bool TxListModel::isRBF(int row) const
...
-   const auto& itTxDet = txDetails_.find(row);
-   if (itTxDet != txDetails_.end()) {
-      if (itTxDet->second.tx.getTxHeight() == UINT32_MAX) {
-         return 0;
-      }
-      return curBlock_ - itTxDet->second.tx.getTxHeight() + 1;
-   }
-   return UINT32_MAX;
-}
-
```
Comment (sergey-chernikov) [inline GUI/QtQuick/TxListModel.cpp:180]: Why was this code removed? Did you check against ZC TXs?

File: GUI/QtQuick/qml/WalletProperties/WalletSeed.qml [hunk 8]
```diff
@@ -47,16 +47,26 @@ ColumnLayout  {
 
     CustomButton {
         id: confirm_but
-        text: qsTr("Copy seed")
+        text: qsTr("Print PDF")
```
Comment (sergey-chernikov) [inline GUI/QtQuick/qml/WalletProperties/WalletSeed.qml]: According to the implementation PDF is not currently printed, but only saved to a file - will it change?

File: GUI/QtQuick/PaperBackupWriter.h (hunk 4 above)
Comment (yzhurko) [reply GUI/QtQuick/PaperBackupWriter.h:27]: No, that's correct sizes for A4 in inches

File: GUI/QtQuick/PaperBackupWriter.h (hunk 5 above)
Comment (yzhurko) [reply GUI/QtQuick/PaperBackupWriter.h]: Removed all comments with this style - these 2 files PaperBackupWriter.* was copy pasted somewhere from common ui with additional small changes.

File: GUI/QtQuick/TxListModel.cpp (hunk 7 above)
Comment (yzhurko) [reply GUI/QtQuick/TxListModel.cpp:180]: Recheck it again - looks like everything works correctly. I'm able to resend transaction again while I have only 0 confirmations.
I removed this piece as it doesn't work correctly - for a new transaction it always returns - 0 confirmations, while actual number is growing.

File: GUI/QtQuick/qml/WalletProperties/WalletSeed.qml (hunk 8 above)
Comment (yzhurko) [reply GUI/QtQuick/qml/WalletProperties/WalletSeed.qml]: Right, renamed button to "Save PDF" for now

**Comment by yzhurko** [comment]
@sergey-chernikov please, have a look into the changes again

File: GUI/QtQuick/QtQuickAdapter.cpp [hunk 9]
```diff
@@ -2592,3 +2593,36 @@ bool QtQuickAdapter::broadcastSignedTX(const QUrl& path)
    pushRequest(user_, userBlockchain_, msg.SerializeAsString());
    return true;
 }
+
+bool QtQuickAdapter::isRequestReadyToSend(QTXSignRequest* request)
+{
+    const auto& walletIds = request->txReq().walletIds;
```
Comment (sergey-chernikov) [inline GUI/QtQuick/QtQuickAdapter.cpp]: Wrong indent size

File: GUI/QtQuick/TxListModel.cpp [hunk 10]
```diff
@@ -115,20 +115,25 @@ QColor TxListModel::dataColor(int row, int col) const
...
 QString TxListModel::walletName(int row) const
 {
-   const auto& itTxDet = txDetails_.find(row);
-   if (itTxDet != txDetails_.end()) {
-      return QString::fromStdString(itTxDet->second.walletName);
-   }
-   return {};
+    const auto& itTxDet = txDetails_.find(row);
+    if (itTxDet != txDetails_.end()) {
+        return QString::fromStdString(itTxDet->second.walletName);
```
Comment (sergey-chernikov) [inline GUI/QtQuick/TxListModel.cpp]: wrong indent

**Review by sergey-chernikov — APPROVED** [review]

**Pull request merged**